apache/superset
# feat(SIP-39): Async query support for charts

### SUMMARY
Adds support for running chart queries in async workers, as proposed by [SIP-39](https://github.com/apache/incubator-superset/issues/9190), under a new `GLOBAL_ASYNC_QUERIES` feature flag.
- Formalizes async event publishing using Redis Streams, with this implementation using a new `/api/v1/async_event/` endpoint for polling.
- Adds async event middleware to `superset-frontend` to poll for events when the feature flag is enabled and there are chart components in a `loading` state.
- Adds a new JWT token cookie (name configurable) to the client for validating async requests.
- This PR affects charts using both the new `/api/v1/chart/data` API and the legacy `/superset/explore_json` endpoint in both dashboards and Explore (behind `GLOBAL_ASYNC_QUERIES` feature flag).
- Abstracts and moves `set_and_log_cache` out of `viz.py`.

Remaining issues:
- Multi-line chart not rendering correctly ([component](https://github.com/apache-superset/superset-ui/blob/master/plugins/legacy-preset-chart-nvd3/src/LineMulti/LineMulti.jsx) will need updating as it does its own data fetching)

TODO in separate follow-up PRs:
- Query cancellation for async queries (see limitations in SIP-39)
- Support async operation for CSV, raw results and sample data
- Add support for [websocket event handling](https://github.com/apache/incubator-superset/pull/11498) on the frontend
- Update SQL Lab async query execution to use new async architecture (backwards compatible)

### TEST PLAN

- Enable the `GLOBAL_ASYNC_QUERIES` feature flag in your test environment and ensure correct chart rendering in both dashboards and Explore
- Ensure synchronous chart rendering with the feature flag disabled functions correctly.

### ADDITIONAL INFORMATION


- [x] Has associated issue: https://github.com/apache/incubator-superset/issues/9190
